src/Hooks/StrictEqualityHooks.php

```diff
@@ -126,33 +156,170 @@ private static function isEqualOrdered(Atomic $first_type, Atomic $second_type):
 
     private static function isNotEqualOrdered(Atomic $first_type, Atomic $second_type): bool
     {
-        if ($first_type instanceof Atomic\TString && $second_type instanceof Atomic\TString) {
-            // oh god, I hate this: https://3v4l.org/O7RXC
-            return true;
+        // identical at the moment
+        return self::isEqualOrdered($first_type, $second_type);
+    }
+
+    private static function isUnionCompatibleType(array $left_type_atomics, array $right_type_atomics, string $expr_class): bool {
+        if ($expr_class === Equal::class) {
+            //This is just a trick to avoid handling every way
+            return self::isUnionEqualOrdered($left_type_atomics, $right_type_atomics) || self::isUnionEqualOrdered($right_type_atomics, $left_type_atomics) ||
+                   self::isUnionStringEqualOrdered($left_type_atomics, $right_type_atomics) || self::isUnionStringEqualOrdered($right_type_atomics, $left_type_atomics);
+        } else {
+            return self::isUnionNotEqualOrdered($left_type_atomics, $right_type_atomics) || self::isUnionNotEqualOrdered($right_type_atomics, $left_type_atomics) ||
+                   self::isUnionStringNotEqualOrdered($left_type_atomics, $right_type_atomics) || self::isUnionStringNotEqualOrdered($right_type_atomics, $left_type_atomics);
         }
+    }
 
-        if ($first_type instanceof Atomic\TInt && $second_type instanceof Atomic\TInt) {
-            return true;
+    private static function isUnionStringEqualOrdered(array $first_types, array $second_types): bool
+    {
+        foreach ($first_types as $atomic_type) {
+            if ($atomic_type instanceof Atomic\TNonFalsyString) {
+                $with_false = true;
+                $with_true = false;
```

Comment:
I feel like every `$with_xxx` should be combined with previous value. Something like `$with_true = $with_true || false`. Otherwise, you loose whatever value was set up before